### PR TITLE
Migrate formatting_html.py into xarray core

### PR DESCRIPTION
**Pull request by eni-awowale**
This PR migrates the `formatting_html.py` module into `xarray/core/formatting_html.py` as part of the on-going effort to merge `xarray-datatree` into `xarray`.

One thing of note is that importing and setting the `OPTIONS` to `"default"` in `datatree/formatting_html.py` ([lines](https://github.com/xarray-contrib/datatree/blob/main/datatree/formatting_html.py#L13%23L15)) were moved into `xarray/core/options.py` on [#L23](https://github.com/pydata/xarray/blob/main/xarray/core/options.py#L23) and [#L49](https://github.com/pydata/xarray/blob/main/xarray/core/options.py#L49). So, I did not add them back to `xarray/core/formatting_html.py`.

- [x]  Completes migration step for `datatree/formating_htmls.py` https://github.com/pydata/xarray/issues/8572
- [x]  Tests added
- [ ]  ~~User visible changes (including notable bug fixes) are documented in whats-new.rst~~
- [ ]  ~~New functions/methods are listed in api.rst~~

### CODE REVIEW

**Comment by eni-awowale** [comment]
Bringing @TomNicholas comment here!
> Thankyou for working on this @eni-awowale ! The main question I have before diving deeper is whether or not you have any time to look into the annoying bug which squeezes the width of the html output - see [xarray-contrib/datatree#91](https://github.com/xarray-contrib/datatree/issues/91)

Yeah, I can take a look at this.

**Comment by eni-awowale** [comment]
@TomNicholas I think I found the fix! I responded in the thread https://github.com/xarray-contrib/datatree/issues/91 and updated in my latest commit. I am a little concerned that the `min-width` and `max-width` are the same?

File: xarray/static/css/style.css [hunk 1]
```diff
@@ -28,7 +28,7 @@ body.vscode-dark {
 
 .xr-wrap {
   display: block !important;
-  min-width: 300px;
+  min-width: 700px;
```
Comment (eni-awowale) [inline xarray/static/css/style.css:31]: Updating the `min-width` to be `700px` seems to solve this bug, however, now it is the same as the `max-width`. I am not sure if this is okay but thought I would point it out.

Comment (TomNicholas) [reply xarray/static/css/style.css:31]: Awesome! Are there other similar values to compare to in that file? Should we just have `min-width`=700px and max-width even bigger, or is the bug from them being different values somehow?

Comment (TomNicholas) [reply xarray/static/css/style.css:31]: I don't really know what I'm looking at with the css here... Pinging @benbovy and  @jsignell as the original authors of the HTML repr 😁 Can either of you tell us how you would handle this?

File: xarray/core/formatting_html.py [hunk 2]
```diff
@@ -341,3 +343,130 @@ def dataset_repr(ds) -> str:
...
+children_section = partial(
+    _mapping_section,
+    name="Groups",
+    details_func=summarize_datatree_children,
+    max_items_collapse=1,
+    expand_option_name="display_expand_groups",
+)
+
+
+def datatree_node_repr(group_title: str, dt: Any) -> str:
```
Comment (owenlittlejohns) [inline xarray/core/formatting_html.py:383]: Looking at the original `formatting_html.py`, it looks like you inherited the `Any` typing in this file. This seems like a good time to tighten them up; I'm guessing they should probably be `DataTree` - I'll comment on the other two that I found in this file.

File: xarray/core/formatting_html.py [hunk 3]
```diff
@@ -341,3 +343,130 @@ def dataset_repr(ds) -> str:
     ]
 
     return _obj_repr(ds, header_components, sections)
+
+
+def summarize_datatree_children(children: Mapping[str, Any]) -> str:
```
Comment (owenlittlejohns) [inline xarray/core/formatting_html.py:348]: Is this `Any` really a `DataTree`? (Not 100% sure, but it seems like a good guess)

File: xarray/core/formatting_html.py [hunk 4]
```diff
@@ -341,3 +343,130 @@ def dataset_repr(ds) -> str:
...
+            "'>",
+            "<ul class='xr-sections'>",
+            r,
+            "</ul>" "</div>",
+            "</div>",
+        ]
+    )
+
+
+def datatree_repr(dt: Any) -> str:
```
Comment (owenlittlejohns) [inline xarray/core/formatting_html.py:470]: Another `Any` that probably should be a `DataTree`.

File: xarray/static/css/style.css (hunk 1 above)
Comment (jsignell) [reply xarray/static/css/style.css:31]: That width controls what the repr looks like when the area that it has available is small. It basically says the width of the repr cannot be less than 700px. This mostly comes up in jupyterlab when you have a split view. Here's the difference:

main:

image

this PR:

Screenshot from 2024-04-12 09-17-25

see how the repr goes off screen? There is a horizontal scroll, so you can still get to the far right of the repr, but it's just a little less beautiful.

It seems like this change is related to https://github.com/xarray-contrib/datatree/issues/91. Is there a reproducible example that I can mess around with?

Comment (jsignell) [reply xarray/static/css/style.css:31]: I found create_test_datatree so I have a reproducer now.

File: xarray/core/formatting_html.py (hunk 2 above)
Comment (eni-awowale) [reply xarray/core/formatting_html.py]: Thanks for noticing this! Just pushed up a new commit with these updates

**Comment by eni-awowale** [comment]
@owenlittlejohns I noticed I am getting some circular import errors from importing `DataTree` in `formatting_html.py`, so I can use it in place of `Any`.

**Comment by TomNicholas** [comment]
> circular import errors from importing DataTree

Shouldn't that be solvable with `if TYPE_CHECKING` and using `"DataTree"` as the type? Because it's only a type hint, you don't need it at runtime.

**Comment by eni-awowale** [comment]
> > circular import errors from importing DataTree
> 
> Shouldn't that be solvable with `if TYPE_CHECKING` and using `"DataTree"` as the type? Because it's only a type hint, you don't need it at runtime.

Ahh yes that solves it! Thanks!

File: xarray/core/formatting_html.py (hunk 3 above)
Comment (owenlittlejohns) [reply xarray/core/formatting_html.py]: Thanks for fixing up the type hints!

File: xarray/static/css/style.css (hunk 1 above)
Comment (jsignell) [reply xarray/static/css/style.css:31]: Alright I think I have something:

```diff
diff --git a/xarray/core/formatting_html.py b/xarray/core/formatting_html.py
index bef0ab03..d2181f31 100644
--- a/xarray/core/formatting_html.py
+++ b/xarray/core/formatting_html.py
@@ -364,7 +364,7 @@ def summarize_datatree_children(children: Mapping[str, Any]) -> str:
 
     return "".join(
         [
-            "<div style='display: inline-grid; grid-template-columns: 100%'>",
+            "<div style='display: inline-grid; grid-template-columns: 100%; grid-column: 1 / -1'>",
             children_html,
             "</div>",
         ]
@@ -438,7 +438,7 @@ def _wrap_datatree_repr(r: str, end: bool = False) -> str:
     height = "100%" if end is False else "1.2em"
     return "".join(
         [
-            "<div style='display: inline-grid;'>",
+            "<div style='display: inline-grid; grid-template-columns: 0px 20px auto; width: 100%;'>",
             "<div style='",
             "grid-column-start: 1;",
             "border-right: 0.2em solid;",
@@ -459,9 +459,8 @@ def _wrap_datatree_repr(r: str, end: bool = False) -> str:
             "<div style='",
             "grid-column-start: 3;",
             "'>",
-            "<ul class='xr-sections'>",
             r,
-            "</ul>" "</div>",
+            "</div>",
             "</div>",
         ]
     )
diff --git a/xarray/static/css/style.css b/xarray/static/css/style.css
index 0079e7d3..e0a51312 100644
--- a/xarray/static/css/style.css
+++ b/xarray/static/css/style.css
@@ -28,7 +28,7 @@ body.vscode-dark {
 
 .xr-wrap {
   display: block !important;
-  min-width: 700px;
+  min-width: 300px;
   max-width: 700px;
 }
 
```

Comment (jsignell) [reply xarray/static/css/style.css:31]: image

Comment (TomNicholas) [reply xarray/static/css/style.css:31]: Amazing! Thank you @jsignell ! @eni-awowale could you include this and maybe stress-test it a little?

Comment (eni-awowale) [reply xarray/static/css/style.css:31]: Thanks for dropping this in @jsignell ! From testing this on a NASA GESDISC dataset with groups I am still seeing the same issue between a `xr.Dataset` and a `datatree.DataTree`. 
Screenshot 2024-04-12 at 5 32 08 PM

Maybe someone else could try testing it on their end?

Comment (owenlittlejohns) [reply xarray/static/css/style.css:31]: I had a quick play around taking this branch and applying the changes @jsignell included above. [Here's a link to the GESDISC data](http://tropomi.gesdisc.eosdis.nasa.gov/test/data//S5P_TROPOMI_Level2/S5P_L2__CLOUD__HiR.2/2023/238/S5P_OFFL_L2__CLOUD__20230826T111842_20230826T130011_30405_03_020500_20230828T031255.nc) mentioned by @eni-awowale.

To me, this looks good. But maybe I'm missing something?

Screenshot 2024-04-17 at 10 44 19 AM

Comment (owenlittlejohns) [reply xarray/static/css/style.css]: I chatted with @eni-awowale and have pushed a commit with the changes from @jsignell to this branch.

Comment (TomNicholas) [reply xarray/static/css/style.css]: That looks good to me too - so is do we think this is resolved now?

Comment (owenlittlejohns) [reply xarray/static/css/style.css]: ...and I broke some tests. Will fix those now.

Comment (owenlittlejohns) [reply xarray/static/css/style.css]: Fixed: [fdf53f](https://github.com/pydata/xarray/pull/8930/commits/fdf53f5d290c1e1e98197ef5046ce07eecb3e87b)

It looks resolved to me, but want to check sure if @eni-awowale is okay with things as she saw different behaviour.

Comment (eni-awowale) [reply xarray/static/css/style.css]: I will check this out as soon as I can!

Comment (eni-awowale) [reply xarray/static/css/style.css]: Just tried it out and it's fixed! I will update this to not be a draft

Comment (owenlittlejohns) [reply xarray/static/css/style.css]: @TomNicholas - I think we're good to go.

**Review by flamingbear — APPROVED** [review]
I like these changes, and I saw the CSS update worked. Nice. Approving even though that doesn't help you.

**Review by TomNicholas — APPROVED** [review]
> Approving even though that doesn't help you.

Explicitly approving it helps me!

Thank you everyone!

**Pull request merged**